File: server/services/promptContextService.js
import { listAnnotations } from './annotationService.js';
import { getComponentRegistry } from './modelMetadataService.js';

const MAX_CONTEXT_COMPONENTS = 80;
const MAX_CONTEXT_ANNOTATIONS = 30;
const MAX_RECENT_ITEMS = 8;

function compactComponent(component) {
  return {
    id: component.id,
    name: component.name,
    meshName: component.meshName,
    materialName: component.materialName,
    editable: component.editable,
    allowedOperations: component.allowedOperations,
    transformBounds: component.transformBounds
  };
}

function compactSceneComponent(component) {
  return {
    id: component.id,
    name: component.name,
    visible: component.visible ?? true,
    material: {
      color: component.material?.color ?? null,
      type: component.material?.type ?? null
    },
    transform: {
      position: component.transform?.position ?? [0, 0, 0],
      rotation: component.transform?.rotation ?? [0, 0, 0],
      scale: component.transform?.scale ?? [1, 1, 1]
    }
  };
}

function compactAnnotation(annotation) {
  return {
    id: annotation.id,
    type: annotation.type,
    status: annotation.status,
    variantId: annotation.variantId,
    partId: annotation.partId,
    targetType: annotation.targetType,
    position: annotation.position,
    normal: annotation.normal,
    screenPosition: annotation.screenPosition,
    points: annotation.points,
    screenPoints: annotation.screenPoints,
    cutPlane: annotation.cutPlane,
    label: annotation.label,
    note: annotation.note,
    updatedAt: annotation.updatedAt
  };
}

function uniqueValues(values) {
  return Array.from(new Set(values.filter(Boolean)));
}

function getCurrentVariant(workspace) {
  const currentVariant = workspace.variantHistory.find((variant) => variant.id === workspace.currentVariantId);
  return {
    id: workspace.currentVariantId,
    label: currentVariant?.label ?? null,
    lastOperations: workspace.lastOperations.slice(-MAX_RECENT_ITEMS)
  };
}

export function buildPromptContext(workspace) {
  const registry = getComponentRegistry(workspace.projectId);
  const activeAnnotations = listAnnotations(workspace.projectId, { status: 'open' });
  const variantScopedAnnotations = activeAnnotations.filter((annotation) => {
    return !annotation.variantId || !workspace.currentVariantId || annotation.variantId === workspace.currentVariantId;
  });
  const components = registry.components.length
    ? registry.components.map(compactComponent)
    : workspace.scene.components.map((component) => ({
        id: component.id,
        name: component.name ?? component.id,
        meshName: component.name ?? component.id,
        materialName: component.material?.type ?? '',
        editable: true,
        allowedOperations: ['colour', 'material', 'visibility', 'scale', 'position', 'rotation'],
        transformBounds: null
      }));

  const selectedComponent = components.find((component) => component.id === workspace.selectedPartId) ?? null;
  const compactAnnotations = variantScopedAnnotations.slice(0, MAX_CONTEXT_ANNOTATIONS).map(compactAnnotation);
  const lineGuides = compactAnnotations.filter((annotation) => annotation.type === 'line');
  const cutGuides = compactAnnotations.filter((annotation) => annotation.type === 'cut_guide');

  return {
    projectId: workspace.projectId,
    workspaceId: workspace.workspaceId,
    modelId: workspace.modelId,
    selectedPartId: workspace.selectedPartId,
    selectedTool: workspace.selectedTool,
    rightPanelMode: workspace.rightPanelMode,
    viewport: workspace.viewport,
    currentVariant: getCurrentVariant(workspace),
    selectedComponent,
    components: components.slice(0, MAX_CONTEXT_COMPONENTS),
    sceneComponents: workspace.scene.components.slice(0, MAX_CONTEXT_COMPONENTS).map(compactSceneComponent),
    annotations: compactAnnotations,
    lineGuides,
    cutGuides,
    allowedOperations: uniqueValues(components.flatMap((component) => component.allowedOperations ?? [])),
    recentPrompts: workspace.promptHistory.slice(-MAX_RECENT_ITEMS).map((entry) => ({
      id: entry.id,
      prompt: entry.prompt,
      createdAt: entry.createdAt
    })),
    constraints: {
      noRawGeometry: true,
      destructiveCutsSupported: false,
      operationOutputOnly: true
    }
  };
}

File: backend/app/services/prompt_context_service.py
from app.schemas.components import ComponentRegistry
from app.schemas.workspace import Workspace

MAX_CONTEXT_COMPONENTS = 80
MAX_RECENT_ITEMS = 8


def build_prompt_context(
    registry: ComponentRegistry,
    *,
    scene_id: str | None = None,
    selected_component_id: str | None = None,
    workspace: Workspace | None = None,
) -> dict:
    selected_id = selected_component_id or (workspace.selected_part_id if workspace else None)
    selected = None
    components = []
    for component in registry.components[:MAX_CONTEXT_COMPONENTS]:
        item = {
            "id": component.id,
            "name": component.name,
            "type": infer_component_type(component),
            "editable": component.editable,
            "meshName": component.mesh_name,
            "allowedOperations": component.allowed_operations,
            "tags": build_component_tags(component),
        }
        components.append(item)
        if selected_id and component.id == selected_id:
            selected = item

    return {
        "scene": {
            "id": scene_id,
            "revisionId": registry.updated_at,
        },
        "workspace": build_workspace_context(workspace),
        "selectedComponent": selected,
        "components": components,
        "sceneComponents": build_scene_components(workspace),
        "allowedOperationTypes": ["setColor", "setVisibility", "setMaterial", "setPosition", "setScale", "setRotation"],
        "constraints": {
            "noRawGeometry": True,
            "operationOutputOnly": True,
        },
    }


def build_workspace_context(workspace: Workspace | None) -> dict | None:
    if not workspace:
        return None
    return {
        "id": workspace.workspace_id,
        "projectId": workspace.project_id,
        "modelId": workspace.model_id,
        "selectedPartId": workspace.selected_part_id,
        "currentVariantId": workspace.current_variant_id,
        "lastOperations": [operation.model_dump(by_alias=True) for operation in workspace.last_operations[-MAX_RECENT_ITEMS:]],
    }


def build_scene_components(workspace: Workspace | None) -> list[dict]:
    if not workspace:
        return []
    return [
        {
            "id": component.get("id"),
            "name": component.get("name"),
            "visible": component.get("visible", True),
            "material": {
                "color": component.get("material", {}).get("color"),
                "type": component.get("material", {}).get("type"),
            },
            "transform": {
                "position": component.get("transform", {}).get("position", [0, 0, 0]),
                "rotation": component.get("transform", {}).get("rotation", [0, 0, 0]),
                "scale": component.get("transform", {}).get("scale", [1, 1, 1]),
            },
        }
        for component in workspace.scene.components[:MAX_CONTEXT_COMPONENTS]
    ]


def infer_component_type(component) -> str:
    for operation in ("annotation", "line", "cut_annotation"):
        if operation in component.allowed_operations and len(component.allowed_operations) == 1:
            return operation
    name = f"{component.id} {component.name or ''} {component.mesh_name}".lower()
    if "camera" in name:
        return "camera"
    if "light" in name:
        return "light"
    if "label" in name:
        return "label"
    if "sign" in name:
        return "sign"
    if "road" in name:
        return "road"
    if "child" in name or "cat" in name:
        return "character"
    return "prop"


def build_component_tags(component) -> list[str]:
    values = {component.id.replace("_", " ")}
    if component.name:
        values.add(component.name.lower())
    if component.mesh_name:
        values.add(component.mesh_name.replace("_", " ").lower())
    values.update(build_alias_tags(component.id, component.name or "", component.mesh_name))
    return sorted(values)


def build_alias_tags(component_id: str, name: str, mesh_name: str) -> set[str]:
    text = f"{component_id} {name} {mesh_name}".lower().replace("_", " ")
    aliases: set[str] = set()
    if "chair seat" in text:
        aliases.add("cushion")
        aliases.add("seat")
    if "chair legs" in text:
        aliases.add("legs")
    if "hat red band" in text or "hat band" in text:
        aliases.add("hat band")
        aliases.add("band")
    if "top hat" in text or "hat top" in text:
        aliases.add("top hat")
        aliases.add("hat")
    return aliases
